language.md

```diff
@@ -7,312 +7,429 @@ the grammar of a synthesis problem, semantics for each production, and constrain
+  ...)
 ```
 
-#### Operators, Leaves, and CHC Bodies
-Leaves are productions with no child terms:
+#### CHC Declarations for Semantic Rules
+
+The semantic rules themselves are SMT-LIB terms expressing a relation between the arguments to the
+function as declared above. They should consist of an `exists`, which allows for the declaration of
+intermediate variables in the CHC, whose inner term is an `and`, whose clauses are either semantic
+relations for child terms or conjuncts for the CHC's constraint.
+
 ```lisp
-(x.Syn (= r x))
+;; CHC for the + operator with children et1: E and et2: E
+;; forall r1, r2. E.Sem(et1, x, y, r1) /\ E.Sem(et2, x, y, r2) /\ r = r1 + r2
+;;             => E.Sem(($+ et1, et2), x, y, r)
+(exists ((r1 Int) (r2 Int)) ; Declare intermediate integer variables r1 and r2
+  (and
+    (E.Sem et1 x y r1) ; Semantic relation for the first child term et1
+    (E.Sem et2 x y r2) ; Semantic relation for the second child term et2
+    (= r (+ r1 r2))))  ; Constraint: the result of + is the sum of the results of et1 and et2
 ```
-This defines a production that is _syntactically_ `x` (named `x.Syn`) and _semantically_ has the CHC body `(= r x)`.
-Together with the non-terminal's semantic relation, it defines the following CHC:
+
+If no intermediate variables are needed, the `exists` may be omitted. Moreover, if there is only one
+clause in the `and` (often the case for leaf terms, where there are no child terms to specify relations
+for), it may be omitted.
+
+```lisp
+;; CHC for the x leaf term, which has no children
+;; r = x => E.Sem($x, x, y, r)
+(= r x) ; Constraint: the result is the input x
 ```
-\forall et,x,y,r. (= r x) => (E.Sem et x y r)
+
+#### Input and Output Annotations
+
+It is possible to annotate an argument to a semantic relation as an "input" or an "output". In semantics
+where program evaluation may be regarded as a transition system, such an annotation asserts that an
+argument represents pre-condition or post-condition data, respectively, for a transition corresponding to
+the term. For example, in an imperative semantics, the pre-condition data are the states of the variables
+prior to a statement's execution, and the post-condition data are the states afterwards.
+
+To annotate the arguments of a semantic relation, we wrap its body declaration with a `!` and append the
+annotations behind it.
+
+```lisp
+;; For the semantic relation E.Sem(t_e, x, y, r), we annotate x and y as inputs and r as an output
+(! (match t_e (<match cases>)) :input (x y) :output (r))
 ```
 
-Operators are the same, except with child terms:
+Note that these annotations are strictly optional, and only serve to assist certain solvers that may be
+able to make use of them.
+
+#### Synthesis Objective
+
+The `synth-fun` command gives a name to a term to be synthesized and specifies its term type. Note that
+the solution to a SemGuS problem is a term and not a function, as imperative semantics must be supported.

```

Comment:
I don't really like the phrasing ", as imperative semantics must be supported." (I know, I wrote that phase a while ago).

I don't know if we need a "because" here, or if it should have the sentiment that we support more than just functions.

Reply:
I'm not really sure that we even need that sentence, since I would think it's pretty clear that we're synthesizing terms from the fact that we give the synthesis target a term type. I just left that sentence in because it was already there, I guess

Reply:
Removing it would work for me.